Max pool grad grad

aam-at opened the pull request:
Adds gpu and cpu kernels for max pool grad. Supports NHWC and NCHW layouts for gpu kernel and only NHWC for cpu kernel. Additionally registers gradient for python `MaxPoolGradGrad` which is just `MaxPoolGrad`. Registers gradient for python `AvgPoolGrad` which is `AvgPool` evaluated on `top_diff`.

Continuation of #6299

aam-at committed:
Adds cpu and gpu kernels for max pooling grad grad

aam-at committed:
Adds python gradient definition for average and max pooling

aam-at committed:
Adds maxpool grad grad to the list of hidden ops

aam-at committed:
Update shape inference function for MaxPoolGradGrad

aam-at committed:
Fixes max pool grad grad cpu kernel

aam-at committed:
Adds grad grad grad for max pool and avg pool

aam-at committed:
Updates gpu kernel to correctly accumulate gradient

aam-at committed:
Adds tests for max pool grad grad (python bindings)

aam-at committed:
Fixes max pool grad grad op error message

aam-at committed:
Updates cpu and gpu kernels (accumulate only single value)

Current implementation propagates only one maximum value in top_diff. To be
compatible with cudnn and the rest of tensorflow api propagate only single value
(specifically should_stop flag is added into inner loop of kernel)

aam-at committed:
Adds gpu grad grad kernel for HCWC data layout

tensorflow-jenkins commented:
Can one of the admins verify this patch?

aam-at commented:
I signed it!

aam-at committed:
Defines pooling ops for all real number types

aam-at committed:
Registers kernels properly

Group kernels and register them using compiler macros. This allows to easily
add new types for kernels.

aam-at committed:
Declares functors for max pooling kernels with support for double

This allows to use templates and conveniently register kernels. Further
refactoring may combine cpu and gpu ops. However, currently cpu does not support
all the variants which are available for gpu.

aam-at committed:
Updates cudnn pooling bindings to support double

Changes to core stream_executor merely duplicate method for float with
different type.

aam-at committed:
Adds test for maxpooling with double

aam-at commented:
I did some refactoring for `MaxPoolingOp` which now uses functors and templates and registers kernels using `TF_CALL_*`. Since currently there more kernels for gpu then cpu, functors are not generalized to devices. Also adding `double` required a minor tweak to `stream_executor` (merely duplicating methods for `double`). I also added python tests with `double` now. This should fix #547

I also updated registered ops to use more general type - `realnumbertype`. Note, since I am not using default type for `T` ops compatibility test fails. Therefore, I have not updated yet `ops.pbtxt`.

I also would like to add `MaxPool3dGradGrad` (currently I added only forward declaration for it).

As this pull request is getting quite big, I would appreciate some comments.

aam-at committed:
Increase margin for second order gradient to 1.5e-2

testMaxPoolGradValidPadding2_1_7_3d on gpu fails with 1e-2 margin (error is
0.102). Increases margin so tests can pass. On cpu tests passes with 1e-2
margin. The differences are related to different numerical precision on gpu and
cpu.

aam-at commented:
I updated 3d kernels. Cpu kernel supports only NHWC and gpu kernel supports both. Currently, python tests for 3d pooling check only NHWC because gpu kernels for pooling and pooling grad only work with this format.

Someone can correct me if I am wrong, but cudnn 5 supports both formats of tensors (NHWC and NCHW). However, both pooling 2d and 3d kernels with cudnn backend always convert tensors from NHWC to NCHW format.

vrv commented:
This is really nice @aam-at: we're waiting for Benoit to do an initial review of this code (he's been on vacation the past week or so).

vrv commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -1102,45 +1112,84 @@ REGISTER_OP("MaxPool3D")
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr("T: realnumbertype")
+    .SetShapeFn(shape_inference::Pool3DShape)
```
small note: this is doing validation that 'orig_input' and 'output' follow Pool3DShape's shape contract, but doesn't validate orig_output and 'grad'

I would recommend augmenting this registration with something like:

```
  .SetShapeFn([](InferenceContext* c) {
    TF_RETURN_IF_ERROR(shape_inference::Pool3DShape(c));
    ShapeHandle unused;
    // Validate 'grad' is the same shape as 'orig_input'
    TF_RETURN_IF_ERROR(c->Merge(c->input(0), c->input(2), &unused));
    // Validate 'orig_output' is same shape as 'output'
    TF_RETURN_IF_ERROR(c->Merge(c->input(1), c->output(0), &unused));
    return Status::OK();
  })
```

or something like that

aam-at replied:
Thanks! I will add it in the next commit.

vrv commented on tensorflow/core/ops/ops.pbtxt:
```diff
@@ -10313,6 +10313,154 @@ op {
   summary: "Computes gradients of the maxpooling function."
 }
 op {
+  name: "MaxPoolGradGrad"
```
Thanks for making these changes, though you don't need to add them in this PR -- they are automatically generated later from the op registrations.

aam-at replied:
Okay. I may drop these changes then. Also, as I noted before, the previous declaration of MaxPool3dGrad didn't use template attribute `T` for types of input and output. So ops compatibility tests will fail. I assume that the correct declaration is to use template attribute.

vrv replied:
Oh, you probably need to read this: https://github.com/tensorflow/tensorflow/blob/754048a0453a04a761e112ae5d99c149eb9910dd/tensorflow/g3doc/how_tos/adding_an_op/index.md#backwards-compatibility

My suggestion would be to declare this as a new type Tinput with a default of DT_FLOAT to preserve backwards compatibility as in the suggestion in the link above.  It sucks to have to declare a new type for this, but I can't think of alternative that is backwards compatible.

aam-at replied:
Yeah, that worked. I assume you meant something like this:
```c++
REGISTER_OP("MaxPool3DGrad")
    .Input("orig_input: Tinput")
    .Input("orig_output: Tinput")
    .Input("grad: T")
    .Output("output: T")
    .Attr("T: numbertype = DT_FLOAT")
    .Attr("Tinput: numbertype = DT_FLOAT");
```
Compatibility test passes now but I assume that original definition contained mistake. Do we want to preserve backward compatibility for this op?

aam-at committed:
Adds 3-rd order gradient for max pooling

aam-at committed:
Remove redundant comment (eigen pooling works only with floats)

aam-at committed:
Update shape function for max pool grad grad

More comprehensive check for input and output shapes for op.

aam-at commented:
I dropped commit changing ops.pbtxt which required rebase.

aam-at committed:
Respect clang google format

aam-at commented:
Any update on reviewing this?

ahundt commented:
@vrv @benoitsteiner max pool grad grad definitely looks like some nice changes, would appreciate the review + merge!

vrv commented:
I'll kick off a test anyway, but will wait for @benoitsteiner to review.

@tensorflow-jenkins test this please

aam-at committed:
Update cudnn pooling loading for double

Reference commit: 191658d54f90ac03c15b339326129cd52d1f56a3

aam-at committed:
Fix function name typo in python

aam-at commented:
The build was broken due to the changes in the upstream (191658d54f90ac03c15b339326129cd52d1f56a3). It is a bit confusing that testing is performed after merging with upstream, though, quite reasonable.

I updated pooling with doubles to call new cudnn loading mechanism. So build should be successful now. @vrv Can you rerun tests again?

vrv commented:
@tensorflow-jenkins test this please

(I don't know the status of our builds right now, so just trying anyway since it's the weekend and the load is low)

vrv commented:
Looks like the sanity check is failing:

=== Sanity check step 3 of 6: do_buildifier (buildifier check) ===

Running do_buildifier on 157 files

tensorflow/core/kernels/BUILD # listsort unsafesort sort:tf_kernel_library.hdrs

buildifier took 0 s

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
2648d2647
<         "pooling_ops_common.h",
2649a2649
>         "pooling_ops_common.h",
Please fix manually or run buildifier <file> to auto-fix.

aam-at commented:
Buildifier requires sorted list of headers and sources. So I fixed that. It's my
first pull request to tensorflow, so I am a bit confused with the requirements
for various tests. Thanks, @vrv!

vrv commented:
@tensorflow-jenkins test this please

@gunan one thing that would be cool would be to have the sanity checks run automatically when users push their PRs -- but then we'd control the rest of the testing.

vrv commented:
@aam-at this is a pretty epic first PR to TensorFlow, so thank you for your patience :)

aam-at commented:
Tests are passing, except for macos. I don't have access to macos and I am not
sure why the build is failing. Thank you for your patience too, @vrv!

vrv commented:
I think it's a infrastructure failure, the fact that the other tests pass suggests this is ready for review!

gunan commented:
I think this can be an issue in build files. But bazel 0.4.3 has a bug in exception propagation, which make us see unhelpful messages. I will upgrade what I can access and rerun tests.

gunan commented:
Jenkins, test this please

aam-at commented:
All tests are passing now. Thanks for the help! I will wait for review now.

vrv commented:
ping for @benoitsteiner for this week.

ahundt commented:
@vrv @benoitsteiner Bumping this item since last post was 10 days ago. :-)

aam-at commented:
@vrv @benoitsteiner any update on this?

vrv commented:
@andydavis1 or @rmlarsen any takers? :)   I am swamped with a lot of other things :(

jhseu commented:
@rmlarsen, do you have time to look at this change?

rmlarsen commented:
I'll take a look a this sometime next week.

rmlarsen commented on tensorflow/core/kernels/maxpooling_op.cc:
```diff
@@ -403,12 +364,260 @@ class MaxPoolingGradOp<Eigen::GpuDevice, T> : public OpKernel {
+    TensorShape input_shape = tensor_in.shape();
+    TensorShape output_shape = tensor_out.shape();
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output));
```
Try:

OP_REQUIRES_OK(context, context->forward_input_or_allocate_output(
{2}, 0, tensor_out.shape(), &output));

rmlarsen commented on tensorflow/core/kernels/maxpooling_op.cc:
```diff
@@ -403,12 +364,260 @@ class MaxPoolingGradOp<Eigen::GpuDevice, T> : public OpKernel {
+    TensorShape output_shape = tensor_out.shape();
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output));
+    if (!context->status().ok()) {
```
you don't need this. The OP_REQUIRES_OK macro tests the status and returns if not OK.

rmlarsen commented on tensorflow/core/kernels/maxpooling_op.cc:
```diff
@@ -403,12 +364,260 @@ class MaxPoolingGradOp<Eigen::GpuDevice, T> : public OpKernel {
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context,
+                   context->allocate_output(0, tensor_out.shape(), &output));
+    if (!context->status().ok()) {
```
you don't need this. The OP_REQUIRES_OK macro tests the status and returns if not OK.

aam-at replied:
Okay. I suspected that but previous implementation of GradOp had an explicit checks. I will remove unnecessary checks throughout `maxpooling` file then.

rmlarsen commented on tensorflow/core/kernels/maxpooling_op_gpu.cu.cc:
```diff
@@ -199,32 +199,111 @@ __global__ void MaxPoolBackward(const int nthreads, const dtype* top_diff,
-  return d.ok();
+template <typename dtype>
+__global__ void MaxPoolGradBackward(const int nthreads, const dtype* top_diff,
+                                    const int64* mask, const int top_offset,
```
"mask" seems a little misleading. Isn't it really argmax within the window?

aam-at replied:
It has the same size as the output `N*C*Hout*Wout` and contains the corresponding index in the input for each element in the output.

rmlarsen replied:
Exactly, so I would not name it mask, but "indices" or "argmax" or something like that.

aam-at replied:
Some kernels were borrowed from caffe, so we should rename `mask` there too. In addition, how should we rename functions like `MaxPoolBackwardNoMask`? In my opinion, the naming may be not the most intuitive (like `max_to_indices`) but it does not introduce too much confusion.

rmlarsen commented on tensorflow/core/kernels/maxpooling_op_gpu.cu.cc:
```diff
@@ -199,32 +199,111 @@ __global__ void MaxPoolBackward(const int nthreads, const dtype* top_diff,
-    float* top_data, int64* mask, const Eigen::GpuDevice& d) {
-  const int kThreadsPerBlock = 1024;
-  const int output_size = batch * channels * pooled_height * pooled_width;
+    const dtype* top_diff, dtype* bottom_diff) {
```
I realize that you didn't start this, but could you please add a comment clarifying what "top" and "bottom" refer to? (I agree that it is preferably to use those terms compared to further overloading "input" and "output".)

Same everywhere.

rmlarsen reviewed (changes requested):
First round of comments.

rmlarsen commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -1102,45 +1112,93 @@ REGISTER_OP("MaxPool3D")
 REGISTER_OP("MaxPool3DGrad")
-    .Input("orig_input: float")
-    .Input("orig_output: float")
+    .Input("orig_input: TInput")
```
Why allow separate types for input and output? The only existing registrations set them both to float:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/pooling_ops_3d.cc#L287
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/pooling_ops_3d.cc#L490

aam-at replied:
It is related to compatibility tests. Previous registration uses `orig_input: float` and `grad: T`. If I try to generalize and set `orig_input: T`, the compatibility test will fail. It was previously discussed here - https://github.com/tensorflow/tensorflow/pull/6664#discussion_r96332231 and I followed @vrv suggestion.

rmlarsen commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -1340,6 +1435,39 @@ argmax: The indices of the maximum values chosen for each output of `max_pool`.
+      return Status::OK();
+    })
+    .Doc(R"doc(
+Computes gradients of the maxpooling backward function.
```
Update comment. Should be "gradients of gradients" or  "second-order" gradients".

aam-at replied:
I updated doc to use your wording "second-order gradients". However, note that C++ code does not compute "final" gradients. GradGradOp will return gradient w.r.t. max pooling gradient. Second-order gradient will be correctly handled in backend gradient implementation, such as python.

rmlarsen commented on tensorflow/python/kernel_tests/pooling_ops_test.py:
```diff
@@ -597,6 +598,38 @@ def _CompareMaxPoolingBk(self, input_shape, output_shape, ksize, strides,
+                                                padding)
+        cpu_val = out_op.eval()
+        self.assertShapeEqual(cpu_val, out_op)
+      if dtype == np.float16:
```
You can simplify this by using assertAllCloseAccordingToType().

rmlarsen commented on tensorflow/python/kernel_tests/pooling_ops_test.py:
```diff
@@ -715,6 +771,83 @@ def _ConstructAndTestGradient(self,
+      input_tensor = constant_op.constant(x, shape=input_sizes, name="input")
+      if pool_func == nn_ops.avg_pool:
+        func_name = "avg_pool"
+        err_margin = 1e-3
```
s/margin/tolerance/

rmlarsen commented on tensorflow/python/kernel_tests/pooling_ops_test.py:
```diff
@@ -715,6 +771,83 @@ def _ConstructAndTestGradient(self,
+                                      x_init_value=None):
+    """Verifies the gradients of gradients of the pooling function.
+
+    Args:
```
cut and paste error?

rmlarsen commented on tensorflow/python/kernel_tests/pooling_ops_test.py:
```diff
@@ -1269,6 +1538,17 @@ def Test(self):
+
 
 if __name__ == "__main__":
   for (name_, input_size_, filter_size_, output_size_, stride_,
```
This is broken. The test names are not unique and therefore only one of the possible tests will get run. This should set a test name that is a concatenation of all the test params. I also recommend using a function like 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/matmul_op_test.py#L35

to add the tests so this will not happen in the future.

aam-at replied:
`GetShrunkInceptionMaxPoolShapes` returns unique name for each parameters test combination, so it should be okay.

rmlarsen commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -3069,6 +3069,41 @@ bool CudnnSupport::DoActivate(Stream* stream,
+  }
+
+  // Alpha is the scaling factor for input.
+  float alpha = 1.0;
```
double?

rmlarsen replied:
The cuDNN documentation states that:

"storage data type for alpha[0], beta[0] is float for HALF and SINGLE tensors, and double
for DOUBLE tensors. These parameters are passed using a host memory pointer"

rmlarsen commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -3138,6 +3173,44 @@ bool CudnnSupport::DoPoolForward(
+  }
+
+  // Alpha is the scaling factor for input.
+  float alpha = 1.0;
```
double?

aam-at committed:
Fix typo mistake

aam-at commented on tensorflow/core/kernels/maxpooling_op.cc:
```diff
@@ -403,12 +364,260 @@ class MaxPoolingGradOp<Eigen::GpuDevice, T> : public OpKernel {
+    PoolParameters params{context,  ksize_,      stride_,
+                          padding_, FORMAT_NHWC, tensor_in.shape()};
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context,
```
It seems that it was added only recently in the upstream. I will do rebase for that.

aam-at committed:
Remove redundant context checks

aam-at committed:
Replace allocate_output calls with forward_input_or_...

aam-at committed:
Use double for alpha and beta with cudnn pooling for doubles

aam-at committed:
Updated doc for MaxPoolingGradGrad

aam-at committed:
Update doc for MaxPoolingGradGrad kernels

aam-at committed:
Rename s/margin/tolerance/ and update doc string

martinwicke commented:
@aam-at what's the status of this?

rmlarsen commented:
@jlebar would you please review the stream executor changes in this PR?

aam-at committed:
Remove unnecessary zero initialization and lunch with cuda config

For MaxPoolGradGrad kernels zero initialization is unnecessary. Therefore,
removing it might save some computation time.

aam-at committed:
Use assertAllCloseAccordingToType

aam-at commented:
@martinwicke It implements MaxPoolingGradGrad kernels for CPU and GPU. Tests are passing on my machine.

rmlarsen commented:
@aam-at Thanks for the updates. Could you please rebase to resolve the conflicts and address the remaining comments.

aam-at committed:
Forward declaration of max pool 3d grad grad and registers gradients

aam-at committed:
Registers 3d pooling with half precision floats

aam-at committed:
Fix confusion with planes/rows/cols in kernel call

aam-at committed:
Update tests for 3d pooling kernels

aam-at commented:
@rmlarsen I have finished rebase. One commit was dropped, since `TensorFormat` handles now 5d tensors. Regarding `mask` parameter, I am not sure if we should rename this parameter because it can introduce further confusion:
1) some kernels were borrowed from caffe
2) not clear how to rename `NoMask` methods.

martinwicke commented:
There are still (or again) conflicts. @rmlarsen if you can resolve the naming issues, we can move this PR along.

rmlarsen commented:
@aam-at OK, let's stick with mask then.

martinwicke commented:
(if this is the last sticking point, make sure to update your review, and I'll switch to just caring about the tests)

rmlarsen commented:
@aam-at please resolve the conflict so we can test this more thoroughly.

rmlarsen reviewed (changes requested):
I would like to run our internal testing before approving this. This requires a clean rebase.

aam-at committed:
Adds operations to tensorflow C++ registry

aam-at committed:
Adds kernels for max pool3d grad grad

aam-at committed:
Buildifier auto-fix for build file

aam-at commented:
@rmlarsen I did rebase again. Currently, there are no conflicts, so you can run the tests.

rmlarsen commented:
@aam-at Thanks!

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@aam-at it looks like there are some broken tests (backwards compatibility and XLA pooling tests) that need attention.

rmlarsen commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -1303,7 +1341,7 @@ output: The gradients for LRN.
 
 REGISTER_OP("MaxPool")
-    .Attr("T: {float, half} = DT_FLOAT")
+    .Attr("T: realnumbertype")
```
Do you need to keep the default value here?

aam-at replied:
The problem was with `MaxPool3dGrad`. We discussed why it is necessary to have two types for input and output, so compatibility tests can pass. I guess during rebase I dropped those changes by accident.

rmlarsen commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -1336,7 +1374,7 @@ REGISTER_OP("MaxPoolGrad")
     .Input("grad: T")
     .Output("output: T")
-    .Attr("T: {float, half} = DT_FLOAT")
+    .Attr("T: realnumbertype")
```
ditto?

aam-at replied:
Fixed.

aam-at committed:
Fix MaxPool3dGrad op definition and default type for maxpool

Previous registration uses orig_input: float and grad: T. To pass compatibility
check two types for the input and output is necessary

aam-at committed:
Fix pooling_ops_3d test for xla compiler

Grad operations for pooling 3d were hidden like pooling 2d grads.

aam-at commented:
@rmlarsen I fixed backward compatibility tests. Xla pooling ops test should be also fixed.

vrv commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@aam-at Excellent. Thank you for this large contribution!

aam-at commented:
@vrv @rmlarsen Thank you for the help!

rmlarsen commented:
@aam-at it appears that the two tests:

_testMaxPoolGradGradSamePadding2_1(data_format, use_gpu)

and 

_testMaxPoolGradGradSamePadding3_1(data_format, use_gpu)

are flaky when use_gpu is True, e.g.:

Traceback (most recent call last):
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 1293, in testMaxPoolGradGrad
    self._testMaxPoolGradGradSamePadding2_1(data_format, use_gpu)
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 1258, in _testMaxPoolGradGradSamePadding2_1
    use_gpu=use_gpu)
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 813, in _ConstructAndTestSecondGradient
    self.assertLess(err, err_tolerance)
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 932, in assertLess
    self.fail(self._formatMessage(msg, standardMsg))
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/testing/pybase/basetest.py", line 1205, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 412, in fail
    raise self.failureException(msg)
AssertionError: 1.99932861328125 not less than 0.01

Perhaps there's a bug for padding is SAME? Do you have any idea what the underlying cause might be?

rmlarsen commented:
The problem was caused by a race due to using buffer forwarding instead of allocation in MaxPoolingGradGradOp(). I'll submit a fix.

aam-at commented:
@rmlarsen It was night in my timezone. Thanks for the fix. I added `forward_input_or_allocated_output` in c1b6aa4b6bbb2e80e228b3ecb9f2877f4e5dc26a. Shouldn't we replace buffer forwarding to allocating output for 3d pooling grad grad too? I am actually not sure why `forward_input_or_allocated_output` creates race conditions.
